validator/nitro_machine.go

```diff
@@ -27,42 +27,36 @@ import (
 	"github.com/ethereum/go-ethereum/log"
 )
 
-type staticMachineData struct {
-	machine    *ArbitratorMachine
-	chanSignal chan struct{}
-	ready      bool
-	err        error
-	once       sync.Once
-}
-
 type NitroMachineConfig struct {
-	RootPath                string // prepends all other paths
+	RootPath                string // a folder with various machines in it
 	ProverBinPath           string
 	ModulePaths             []string
 	InitialMachineCachePath string
 }
 
-var StaticNitroMachineConfig = NitroMachineConfig{
-	RootPath:                "./target/machine/",
+var DefaultNitroMachineConfig = NitroMachineConfig{
+	RootPath:                "./target/machines/",
 	ProverBinPath:           "replay.wasm",
 	ModulePaths:             []string{"soft-float.wasm", "wasi_stub.wasm", "go_stub.wasm", "host_io.wasm", "brotli.wasm"},
 	InitialMachineCachePath: "./target/etc/initial-machine-cache",
 }
 
-var zeroStepMachine staticMachineData
-var hostIoMachine staticMachineData
-
 func init() {
 	_, thisfile, _, _ := runtime.Caller(0)
 	projectDir := filepath.Dir(filepath.Dir(thisfile))
-	StaticNitroMachineConfig.RootPath = filepath.Join(projectDir, "target/machine")
+	DefaultNitroMachineConfig.RootPath = filepath.Join(projectDir, "target", "machines")
+}
 
-	zeroStepMachine.chanSignal = make(chan struct{})
-	hostIoMachine.chanSignal = make(chan struct{})
+func (c NitroMachineConfig) getMachinePath(moduleRoot common.Hash) string {
+	if moduleRoot == (common.Hash{}) {
+		return filepath.Join(c.RootPath, "latest")
+	} else {
+		return filepath.Join(c.RootPath, moduleRoot.String())
+	}
 }
 
-func ReadWasmModuleRoot() (common.Hash, error) {
-	fileToRead := path.Join(StaticNitroMachineConfig.RootPath, "module_root")
+func (c NitroMachineConfig) ReadWasmModuleRoot(requestingModuleRoot common.Hash) (common.Hash, error) {
```

Comment:
I've gone with the former. I was initially thinking reading the module root file would be a nice sanity check to prevent accidentally renaming/symlinking the wrong directory, but it'll recompute the root later on anyways